test/archive.cpp

```diff
@@ -280,6 +280,123 @@ TEST(ZimArchive, openSplitZimArchive)
+    ASSERT_EQ(REF_IT->isRedirect(), test_entry.isRedirect());                         \
+    if (REF_IT->isRedirect()) {                                                       \
+      ASSERT_EQ(REF_IT->getRedirectEntryIndex(), test_entry.getRedirectEntryIndex()); \
+    }                                                                                 \
+    auto ref_item = REF_IT->getItem(true);                                            \
+    auto test_item = test_entry.getItem(true);                                        \
+    ASSERT_EQ(ref_item.getClusterIndex(), test_item.getClusterIndex());               \
+    ASSERT_EQ(ref_item.getBlobIndex(), test_item.getBlobIndex());                     \
+    ASSERT_EQ(ref_item.getData(), test_item.getData());                               \
+  }
+
+TEST(ZimArchive, cacheDontImpactReading)
+{
+  const TestCacheConfig cache_configs[] = {
+    {0, 0, 0},
+    {1, 1, 1},
+    {2, 2, 2},
+    {10, 10, 10},
+    {1000, 2000, 1000},
+    {0, 2000, 1000},
+    {1000, 0, 1000},
+    {1000, 2000, 0},
+    {1, 2000, 1000},
+    {1000, 1, 1000},
+    {1000, 2000, 1},
+  };
+
+  for (auto& testfile: getDataFilePath("small.zim")) {
+    auto ref_archive = zim::Archive(testfile.path);
+
+    for (auto cache_config: cache_configs) {
+      auto test_archive = zim::Archive(testfile.path);
+      test_archive.set_dirent_cache_max_size(cache_config.dirent_cache_size);
+      test_archive.set_dirent_lookup_cache_max_size(cache_config.dirent_lookup_cache_size);
+      test_archive.set_cluster_cache_max_size(cache_config.cluster_cache_size);
+
+      EXPECT_EQ(test_archive.get_dirent_cache_max_size(), cache_config.dirent_cache_size);
+      EXPECT_EQ(test_archive.get_dirent_lookup_cache_max_size(), cache_config.dirent_lookup_cache_size);
+      EXPECT_EQ(test_archive.get_cluster_cache_max_size(), cache_config.cluster_cache_size);
+
+      ASSERT_ARCHIVE_EQUIVALENT(ref_archive, test_archive)
+    }
+  }
+}
+
+
+TEST(ZimArchive, cacheChange)
+{
+  for (auto& testfile: getDataFilePath("wikibooks_be_all_nopic_2017-02.zim")) {
+    std::cout << testfile.path << std::endl;
+    auto ref_archive = zim::Archive(testfile.path);
+    auto archive = zim::Archive(testfile.path);
+
+    archive.set_dirent_cache_max_size(30);
+    archive.set_cluster_cache_max_size(5);
+
+    auto range = ref_archive.iterEfficient();
+    auto ref_it = range.begin();
+    ASSERT_ARCHIVE_EQUIVALENT_IT_LIMIT(ref_it, range.end(), archive, 50)
+    EXPECT_EQ(archive.get_dirent_cache_current_size(), 30);
+    EXPECT_EQ(archive.get_cluster_cache_current_size(), 2); // Only 2 clusters in the file
+
+    // Reduce cache size
+    archive.set_dirent_cache_max_size(10);
+    archive.set_cluster_cache_max_size(1);
+
+    EXPECT_EQ(archive.get_dirent_cache_current_size(), 10);
+    EXPECT_EQ(archive.get_cluster_cache_current_size(), 1);
+
+    ASSERT_ARCHIVE_EQUIVALENT_IT_LIMIT(ref_it, range.end(), archive, 50)
```

Comment:
This fragment of the test scenario is somewhat obscure and may be misleading. The `ASSERT_ARCHIVE_EQUIVALENT_IT_LIMIT()` macro has a side effect - it advances its iterator argument. Therefore the second invocation of that macro on the same iterator object continues the iteration (or performs no iteration at all if the end has already been reached before that) and it's not obvious if that is intentional or not. Please add comments explaining what is being tested or change the test accordingly.

Reply:
Yes, it is intentionnal, we want to change the cache size *as we are iterating on it*.
I will add a comment.